WIP browserpod rust blog post (#448)
leaningtech/labs: sites/labs/src/content/blog/browserpod-rust.mdx

---

> title: "Rust support in BrowserPod"

Reviewer: Beyond WASI: Running any Rust (what is the right capitalization) application in the browser with BrowserPod 3.0

description: |
  BrowserPod can now run real Rust programs in the browser, compiled from unmodified source with stock rustc and cargo.
authors:
  - yuri

> pubDate: "August 12 2026"

Reviewer: At this point we should target Thursday the 13th

heroImage: "./banners/BrowserPod_2.0-hero.webp"
tags:
  - BrowserPod
categories:
  - "Release"
---

import BrowserPodProvider from "@leaningtech/astro-theme/components/blog/BrowserPodProvider.svelte";
import BrowserPodTerminal from "@leaningtech/astro-theme/components/blog/BrowserPodTerminal.svelte";

{/* TODO: create the demo project under sites/labs/public/blog/bp-rust (manifest.txt + .scripts/*.sh) before publishing */}

<BrowserPodProvider
	client:only="svelte"
	ctxId="bp-rust-demo"
	version="2.16.0"
	projectSource={{ type: "local", path: "bp-rust" }}
	apiKey="bp1_2cc277b71b4bdd4850b0ee9bd117b2354e9be99d6a5b1cf3379f8135c84a8a5a"
/>

> Today we're announcing **Rust support in BrowserPod**. You can compile a Rust project with `cargo`, targeting BrowserPod, and run the resulting binary in the Browser. The program can access the filesystem, make network requests, run subprocesses, ... all without changing a single line of code!

Reviewer: ```suggestion
We are releasing today BrowserPod 3.0, with full support for running any Rust application in the browser and countless bug fixes for Node.js and Python support.

Our Rust support goes way beyond what can be normally achieve with the existing WASI target, but it's nevertheless based on a clean integrated Rust toolchain. You can compile a Rust project with `cargo` and run the resulting binary in the Browser. The program will be able to access the filesystem, make network requests, run subprocesses and interact with concurrently running applications. And all of this without changing a single line of code!
```

Reviewer: We can also insert here the story about the maintainer of Yarn asking for Rust support and how we could get something to work in a few weeks. Also that this has delayed python that will be our next priority.

In the demo below, you can see the preview version of Yarn 6 (written in Rust) installing a NPM project:

```
<BrowserPodTerminal
	client:only="svelte"
	ctxId="bp-rust-demo"
	description="Yarn 6 running in the browser"
	autoStart={true}
	height="18rem"
	terminalTabs={[
		{
			id: "rg-demo",
			label: "ripgrep",
			echo: false,
			command: "/home/user/yarn",
			args: ["install"],
		},
	]}
/>
```

## What is BrowserPod

> BrowserPod is an in-browser code sandbox. It runs Linux applications compiled to WebAssembly. It aims to implement the full Linux syscall interface, allowing existing applications to be compiled with little to no code change, and multiple applications to interact with each other.

Reviewer: ```suggestion
BrowserPod is an in-browser code sandbox. It's goal is to make it possible to run any Linux application in modern browsers by compiling the application to WebAssembly and implementing the full Linux syscall interface.
```

> It supports outbound internet access for downloading packages or calling APIs, and inbound connections for exposing local development servers.

Reviewer: ```suggestion
It provides a virtualized, locally persistent and efficient filesystem, outbound internet access for downloading packages or calling APIs, and inbound connections for exposing local development servers. BrowserPod supports real parallelism by running each thread or process on an independent Worker, while providing a consistent view of the system to all the running programs. For all purposes it can be considered a OS kernel for the Web platform, implemented in WebAssembly.
```

> That makes it useful for safe in-browser agentic code execution, web-based IDEs and development environments, interactive docs and live demos, educational platforms, and other applications that benefit from sandboxed execution inside a web app.

Reviewer: ```suggestion
This set of features make BrowserPod uniquely suited for safe in-browser agentic code execution, web-based IDEs and development environments, interactive docs and live demos, educational platforms, and other applications that benefit from sandboxed execution inside a web app.
```

## What does it mean to "support" Rust

> When adding support for Node.js in BrowserPod, we had to compile the `node` program for the BrowserPod target (using our C++ compiler, Cheerp), and then our users can simply copy their JS project in a Pod, and run it in the Browser directly, via `node`.

Reviewer: ```suggestion
In previous releases of BrowserPod we focused on Node.js. Our users did not have to worry how to compile for the BrowserPod target, since we provide a pre-compiled Node.js build ourselves, using our C++ compiler, Cheerp.

Rust programs behave very differently, since they need to be compiled Ahead-Of-Time...
```

- `wasm32-wasip1` / `wasm32-wasip2` / `wasm32-wasip3`: These are the WASI targets. They can run in principle in any environment, although their main use is for WebAssembly outside of the Browser. As such, they don't assume the presence of JS, and they provide applications with I/O facilities like a filesystem and networking.
- `wasm32-wali-linux-musl`: This target is intended to run existing Linux applications in a Wasm runtime, by providing the x86-64 linux syscall interface as imports. Its main use case is outside of the Browser.

## Why not just pick WASI?

Our original plan was to implement a WASI (WebAssembly System Interface) layer on top of BrowserPod.

This way we could run any program that targets it, whether it's written in Rust, Go, C, or anything else.

But when trying to compile existing Rust programs to WASI, we quickly realized that it's not just a matter of switching build target: [...]

> ## Why not just pick WALI?

Reviewer: I think this concept should go towards the end. As part of the "Future work" we could say that we will either upstream the support in the Rust compiler or adopt WALI and use the existing target. I think that mentioning CheerpX here in the middle introduces confusion

## Making our own Rust target

Defining a custom Rust target is surprisingly easy: you define your target's properties
in a .json file:

```json
{
  "llvm-target": "wasm32-unknown-unknown",
  "target-pointer-width": 32,
  "target-c-int-width": 32,
  "data-layout": "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20",
  "arch": "wasm64",
  "is-like-wasm": true,
  "target-family": ["unix"],
  "os": "linux",
  "env": "musl",
  "vendor": "browserpod",

  "linker-flavor": "wasm-ld",
  "linker": "rust-lld",
  ...
}
```

And then pass the path of this file as the `--taget` argument. A nightly compiler is needed.

We also need to support our target in the Rust standard library for our combination of arch and os.
This is also relatively straightforward: we want conditional compilation to select the appropriate syscalls
(in our case, they are the same as the x86 ones), and a few smaller details (unwinding behavior, malloc alignment, enabling the Wasm intrinsics, ...).

And finally, we need to override a couple of foundational libraries: `libc` and `linux-raw-sys`.

> `libc` is a requirement for the standard library itself, but many crates also depend on it directly.

Reviewer: Can we summarize these in a table like we do for the projects that work later?

/>

Today we are releasing BrowserPod 3.0, with full support for running any Rust
application in the browser, many fixes for Node.js, and initial Python support.

Our Rust support goes beyond what can normally be achieved with the existing Wasm
targets in terms of standard library features and third-party crate compatibility.
Programs can access the filesystem, make network requests, run subprocesses, and interact
with concurrently running applications. All of this **without changing a single line of code**!

In the demo below, you can see the preview version of [Yarn 6](https://v6.yarnpkg.com/concepts/yarn-6) (written in Rust) installing an NPM project:

<BrowserPodTerminal
	client:only="svelte"
	ctxId="bp-rust-demo"
	description="Yarn 6 running in the browser"
	autoStart={true}
	height="18rem"
	terminalTabs={[
		{
			id: "rg-demo",
			label: "ripgrep",
			echo: false,
			command: "bash",
			args: [],
		},
	]}
/>

## What is BrowserPod

BrowserPod is an in-browser code sandbox. Its goal is to make it possible to run _any_ Linux application in modern browsers by compiling the application to WebAssembly and providing the full Linux syscall interface.

> It provides a virtualized, locally persistent, and efficient filesystem, outbound internet access for downloading packages or calling APIs, and inbound connections for exposing local development servers. BrowserPod supports real parallelism by running each thread or process on an independent Worker, while providing a consistent view of the system to all the running programs. For all purposes, it can be considered **an OS kernel for the Web platform**, implemented in WebAssembly.

Reviewer: ```suggestion
It provides an efficient locally persistent virtual filesystem, outbound internet access for downloading packages or calling APIs, and inbound connections for exposing local development servers. BrowserPod supports real parallelism by running each thread or process on an independent Worker, while providing a consistent view of the system to all the running programs. For all purposes, it can be considered **an OS kernel for the Web platform**, implemented in WebAssembly.
```

Readability fix to avoid multiple command separated lists

This set of features makes BrowserPod uniquely suited for safe in-browser agentic code execution, web-based IDEs and development environments, interactive docs and live demos, educational platforms, and other applications that benefit from sandboxed execution inside a web app.

## Why Rust

[...]

## What does it mean to "support" Rust

In previous releases of BrowserPod we focused on Node.js. Our users did not have to worry about how to compile for the BrowserPod target, since we provide a pre-compiled Node.js build ourselves, using our C++ compiler, [Cheerp](https://cheerp.io).

Rust programs behave very differently, since they need to be compiled ahead of time by the Rust compiler, `rustc`.

> We could in principle compile `rustc` itself for BrowserPod, and allow users to compile their Rust programs directly in the browser, and we might do so in the future.

Reviewer: Let's swap these 2 sentences.

1. Users need to compile the program offline and add the binary
2. Note that the rustc compiler is written in rust and so, in principle....

Author: I am not sure I understand the end result that you have in mind. could you write it down?

Reviewer: ```suggestion
Rust programs behave very differently, since they need to be compiled ahead of time using the Rust compiler, `rustc`. As things stand today the user needs to compile the program offline and add the resulting binary to the Pod. Interestingly, `rustc` itself is written in Rust and we could, in principle, allow users to compile their Rust programs directly in the browser. We need a few additional features to achieve this objective, but it will happen in the near future.
```

For now, the intended workflow is to compile the program offline and add the resulting binary to the Pod.

> So _"supporting Rust"_ means that we provide a Rust toolchain that can produce Wasm binaries that can run in a BrowserPod Pod.

Reviewer: ```suggestion
Effectively _"supporting Rust"_ boils down to providing a Rust toolchain that can produce Wasm binaries for the BrowserPod target.
```

## Existing Rust WebAssembly targets

> Rust already supports a variety of Wasm targets, namely:

Reviewer: ```suggestion
Fortunately we could start from solid ground. [RFC] Rust already supports a variety of Wasm targets, namely:
```

- [`wasm32-unknown-unknown`](https://doc.rust-lang.org/rustc/platform-support/wasm32-unknown-unknown.html): This target is the most barebones and platform agnostic. By itself, it makes no assumptions about the environment it runs in, but crates like [`wasm-bindgen`](https://wasm-bindgen.github.io/wasm-bindgen/) and [`web-sys`](https://docs.rs/web-sys) make it possible to communicate with the JS and Web environment. This target makes sense if you are deliberately targeting the Web from the start. Many standard library features are missing.

> - [`wasm32-unknown-emscripten`](https://doc.rust-lang.org/rustc/platform-support/wasm32-unknown-emscripten.html): This target is intended for standalone programs targeting the Web. It can directly interact with JavaScript code, and already wraps many Web APIs for ease of use. It also emulates some POSIX APIs to an extent, but many Rust standard library features are not implemented, and in general supporting this target requires extensive rewriting.

Reviewer: ```suggestion
- [`wasm32-unknown-emscripten`](https://doc.rust-lang.org/rustc/platform-support/wasm32-unknown-emscripten.html): This target is intended for standalone programs targeting the Web. It can directly interact with JavaScript code, and already wraps many Web APIs for ease of use. It also emulates some POSIX APIs to an extent, but many Rust standard library features are not implemented. In general supporting this target requires extensive rewriting.
```

> - [`wasm32-wasip1`](https://doc.rust-lang.org/rustc/platform-support/wasm32-wasip1.html) / [`wasm32-wasip2`](https://doc.rust-lang.org/rustc/platform-support/wasm32-wasip2.html) / [`wasm32-wasip3`](https://doc.rust-lang.org/rustc/platform-support/wasm32-wasip3.html): These are the [WASI](https://wasi.dev) targets. They can run in principle in any environment, although their main use is for WebAssembly outside of the browser. As such, they don't assume the presence of JS, and they provide applications with I/O facilities like a filesystem and networking.

Reviewer: ```suggestion
- [`wasm32-wasip1`](https://doc.rust-lang.org/rustc/platform-support/wasm32-wasip1.html) / [`wasm32-wasip2`](https://doc.rust-lang.org/rustc/platform-support/wasm32-wasip2.html) / [`wasm32-wasip3`](https://doc.rust-lang.org/rustc/platform-support/wasm32-wasip3.html): These are the [WASI](https://wasi.dev) targets. They can run in principle in any environment, although their main use is for WebAssembly outside of the browser. As such, they don't assume the presence of JS, but [RFC] they provide applications with I/O facilities like a filesystem and networking.
```

> - [`wasm32-wali-linux-musl`](https://doc.rust-lang.org/rustc/platform-support/wasm32-wali-linux.html): This target is intended to run existing Linux applications in a Wasm runtime, by providing the x86-64 linux syscall interface as imports. Its main use case is outside of the browser.

Reviewer: Should we add a note that this is experimental and/or academic at the moment?

## Why not just pick WASI?

> Our original plan was to implement a WASI (WebAssembly System Interface) layer on top of BrowserPod.

Reviewer: Let's remove a newline here

> This way we could run any program that targets it, whether it's written in Rust, Go, C, or anything else.

Reviewer: ```suggestion
This solution would have any program targeting WASI immediately compatible with BrowserPod, whether it's written in Rust, Go, C, or anything else.
```

- Removal of thread usage: `wasm32-wasip3` has some limited cooperative threading support (without real parallelism), but it's very new and supported by neither the standard library nor [tokio](https://tokio.rs).
- Acceptance of reduced functionality: `yarn` relies on external programs for some functionality: in particular `git` (to fetch git dependencies) and `node` (to run lifecycle scripts) are required for basic operations.

BrowserPod already has a solution to these problems, so we decided to skip the middleman
and implement our own Rust target: `wasm32-browserpod-linux-musl`.

## Making our own Rust target

Defining a [custom Rust target](https://doc.rust-lang.org/rustc/targets/custom.html) is _surprisingly easy_: you define your target's properties
in a .json file:

```json
{
  "llvm-target": "wasm32-unknown-unknown",
  "target-pointer-width": 32,
  "target-c-int-width": 32,
  "data-layout": "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20",
  "arch": "wasm64",
  "is-like-wasm": true,
  "target-family": ["unix"],
  "os": "linux",
  "env": "musl",
  "vendor": "browserpod",

  "linker-flavor": "wasm-ld",
  "linker": "rust-lld",
  ...
}
```

> And then pass the path of this file as the `--target` argument. A nightly compiler is needed.

Reviewer: ```suggestion
And then pass the path of this file as the `--target` argument. A nightly build of `rustc` is needed, since official releases don't support off-tree targets [?RFC]
```

> We also need to support our target in the Rust standard library for our combination of arch and os.

Reviewer: I would advise to condense this section in a single paragraph. We can mention the components but without the details. Also, are our changes public (they should be, if so we can point to the repos)

Author: not public. they are just .patch files in lt-nix right now

(in our case, they are the same as the x86 ones), and a few smaller details (unwinding behavior, malloc alignment, enabling the Wasm intrinsics, ...).

And finally, we need to override a couple of foundational libraries:

[`libc`](https://docs.rs/libc) is a requirement for the standard library itself, but many crates also depend on it directly.
We mainly need to link our own [musl](https://musl.libc.org) libc build, and to define some constants like the sizes of C types.

[`linux-raw-sys`](https://docs.rs/linux-raw-sys) is not strictly needed, but it's commonly found in the dependency tree of many projects, and
we need to teach it about our target and the syscalls we expose (again, mostly copying from the existing x86 code paths).

A number of options to `cargo` and `rustc` are needed to tie everything together (unstable feature flags, the sysroot with the C dependencies, the cargo overrides for `libc` and `linux-raw-sys`, ...); we provide wrapper scripts for them, so the user can compile with a simple `cargo build --target wasm32-browserpod-linux-musl`.

## The hacks we did along the way

> In an ideal world, this would be all we need.

Reviewer: Cleanup excessive newlines here

If you look closer at the json snippet of our target definition above, you will notice two interesting things:

- `target_family: ["unix"]`: originally, we had `["unix", "wasm"]`, but many crates use the wasm target family to mean "standalone web build", "WASI", "no filesystem", or similar. Dropping "wasm" seems to be _completely harmless_.

- `arch: "wasm64"`: this is even more surprising, but unfortunately also necessary. `arch="wasm32"` is often used to detect standalone web targets (for example, [`reqwest`](https://docs.rs/reqwest) will replace the whole http client with a `fetch()` implementation). You would think that something would break because of this, but important info like the size of pointers is defined separately, and the actual target passed to the LLVM backend is still `wasm32`, so it all works out!

Of course, we would love to get rid of these hacks, but the proper solution requires awareness in the ecosystem that fully-featured Wasm targets exist, which will take time (but we have a [plan](#future-plans)).

## The payoff

> So, what can we build and run for our new target? Did we reach our goal of (almost) zero code changes required?

Reviewer: Let's avoid the rethorical question here. Let's state that we succeeded with our goal. We can also say that robustness of the solution shocked yourself as well (I think it's true)

Author: ## The payoff

The original goal was to run `yarn`, but since our approach was completely general, we found that
a lot of programs _just work_.

Here is a (non-exaustive) list of popular programs that I tried:

?

Reviewer: ... that I tried, but we expect most applications to work just as well.

Here is a list of the programs that I tried:

| Project                                                     | Rust SLOCs | Changes                          | Status    |
| ----------------------------------------------------------- | ---------- | -------------------------------- | --------- |
| [burntsushi/ripgrep](https://github.com/burntsushi/ripgrep) | ~40k       | **no changes**                   | _working_ |
| [yarnpkg/zpm](https://github.com/yarnpkg/zpm)               | ~50k       | **no changes**                   | _working_ |
| [starship/starship](https://github.com/starship/starship)   | ~50k       | **no changes**                   | _working_ |
| [jj-vcs/jj](https://github.com/jj-vcs/jj)                   | ~200k      | **no changes**                   | _working_ |
| [openai/codex](https://github.com/openai/codex)             | ~1.25M     | disabled some sandboxing options | _working_ |

HERE ANOTHER DEMO with a few of these inside

VIDEO of codex

## Using it

> Rust has _one_ blessed build system, which makes it much easier to provide a good out-of-the-box developer experience compared to, say, C/C++.

Reviewer: _one single_?

Reviewer: out-of-the-box _crosscompiling_ developer experience?

Author: Rust has a single blessed package manager and build system: `cargo`. This makes it much easier for us to provide a good out-of-the-box developer experience compared to, say, C/C++.

?

in particular I should really say "package manager" too, and that it's "for us" easy to provide...

> Rust also has an official way to obtain and manage multiple toolchains: **[rustup](https://rustup.rs)**. We can't be distributed _through_ rustup yet, because we are not an officially supported target, but we can install into its directories, so you can switch a single project over to our toolchain in the usual way:

Reviewer: ```suggestion
Rust also has an official way to obtain and manage multiple toolchains: **[rustup](https://rustup.rs)**. We can't be distributed _through_ rustup yet, because we are not an officially supported target, but we can install into its directories, so you can enable our toolchain for a single project in the usual way:
```

```bash
# manually install our toolchain (one day this could be via rustup too)
curl https://rt.browserpod.io/3.0.0/rust/install.sh | bash

# override the toolchain for the current project with rustup
rustup override set browserpod-3.0.0

# build
cargo build --target wasm32-browserpod-linux-musl

# uninstall via rustup
rustup toolchain uninstall browserpod-3.0.0
```

## Try it now

Head to [console.browserpod.io](https://console.browserpod.io) to get an API key. You can find the documentation on [browserpod.io/docs](https://browserpod.io/docs).

If you have a Rust project you'd like to see running in a browser tab — especially one you expect to break — we'd genuinely like to hear about it. Come find us on [Discord](https://discord.leaningtech.com).

## Future plans

I explained why the existing Wasm targets were not suitable for us, but I glossed
over `wasm32-wali-linux-musl`.

> The goals of [WALI (WebAssembly Linux Interface)](https://arxiv.org/abs/2312.03858) are actually very aligned with those of BrowserPod: compile Linux applications _unmodified_ to Wasm. The main difference really is that BrowserPod targets the x86 version of the Linux syscalls, while WALI targets the x86-64 version. The choice for BrowserPod comes from the fact that its syscall implementations are shared with [CheerpX](https://cheerpx.io), which is a virtual machine for x86 Linux binaries (you may know it from [WebVM](https://webvm.io)).

Reviewer: Let's mention "outside of the browser" here, otherwise it seems WALI has already solved the problem we work on

Author: Although its main use case is outside of the browser, the main goal of [WALI (WebAssembly Linux Interface)](https://arxiv.org/abs/2312.03858) is actually very aligned with that of BrowserPod: compile Linux applications _unmodified_ to Wasm.

Author: I don't like the "main" repetition though

Reviewer: main -> primary or "main goal" just "goal"

Author: or maybe "although it was conceived outside of the browser"?

Reviewer: I think we should make it clear that WALI has nothing to do with running things in the browser. Maybe we should say the intended target are native Linux systems.

		},
	]}
/>

## What is BrowserPod

BrowserPod is an in-browser code sandbox. Its goal is to make it possible to run _any_ Linux application in modern browsers by compiling the application to WebAssembly and providing the full Linux syscall interface.

It provides an efficient, locally persistent virtual filesystem, outbound internet access for downloading packages or calling APIs, and inbound connections for exposing local development servers. BrowserPod supports real parallelism by running each thread or process on an independent Worker, while providing a consistent view of the system to all the running programs. For all purposes, it can be considered **an OS kernel for the Web platform**, implemented in WebAssembly.

This set of features makes BrowserPod uniquely suited for safe in-browser agentic code execution, web-based IDEs and development environments, interactive docs and live demos, educational platforms, and other applications that benefit from sandboxed execution inside a web app.

## Why Rust now?

BrowserPod's ambition is to run an entire Linux userspace in the browser. For the most part, this used to mean compiling a bunch of C/C++ projects.

> Interpreters and runtimes for dynamic languages are also mostly written in C/C++, and so having a C/C++ toolchain (in our case, [Cheerp](https://cheerp.io)) would get you very far.

Reviewer: ```suggestion
Most dynamic languages, such as Python, JavaScript or Ruby run on top of interpreters and runtimes that are also, most usually, written in C/C++, and so having a C/C++ toolchain (in our case, [Cheerp](https://cheerp.io)) would get you very far.
```

Nowadays, this is less and less true. Popular languages like Rust and Go are compiled ahead of time like C/C++, and have their own toolchains.

> Many build tools for Node.js in particular are being written (or rewritten) in Rust.

Reviewer: Remove the newline before this sentence

> For these reasons, Rust was always part of our roadmap for BrowserPod, but we decided to prioritize it over Python and Ruby thanks to a concrete use case from a member of our [community](https://discord.leaningtech.com) on Discord: the upcoming [Yarn 6](https://v6.yarnpkg.com/concepts/yarn-6).

Reviewer: ```suggestion
For these reasons, Rust was always part of our roadmap for BrowserPod, but we decided to prioritize it over Python and Ruby thanks to a concrete use case from a member of our community on [Discord](https://discord.leaningtech.com).
```

> The maintainers would like to have an interactive documentation page, featuring a real yarn build.

Reviewer: ```suggestion
One maintainer of the Yarn package manager expressed interested in building an interactive documentation page, featuring a real yarn build, and running on BrowserPod. But contrarily to previous versions, which were JavaScript, the upcoming [Yarn 6](https://v6.yarnpkg.com/concepts/yarn-6) release is built in Rust.
```

Used "One maintainer" to make sure to stay truthful.

> This is one of the use cases where we think that BrowserPod can really shine, and so we started working on it right away!

Reviewer: ```suggestion
This is one of the use cases where we think that BrowserPod can really shine, and so we started working on it right away! It took us a (two?) weeks to get to a first prototype and we could immediately see the potential, but there were more moving parts that we had expected.
```

## What does it mean to "support" Rust

> In previous releases of BrowserPod we focused on Node.js. Our users did not have to worry about how to compile for the BrowserPod target, since we provide a pre-compiled Node.js build ourselves, using our C++ compiler, [Cheerp](https://cheerp.io).

Reviewer: ```suggestion
In previous releases of BrowserPod we focused on Node.js. Our users did not have to worry about compiling for the BrowserPod target, since we provide a pre-compiled Node.js build ourselves.
```

We mention Cheerp already above, it's enough

> Rust programs behave very differently, since they need to be compiled ahead of time using the Rust compiler, `rustc`. As things stand today, the user needs to compile the program offline and add the resulting binary to the Pod. Interestingly, `rustc` itself is written in Rust, and we could, in principle, allow users to compile their Rust programs directly in the browser. We need a few additional features to achieve this objective, but it will happen in the near future.

Reviewer: ```suggestion
Rust programs behave very differently, since they need to be compiled ahead of time using `rustc`. As things stand today, the user needs to compile the program offline and add the resulting binary to the Pod. Interestingly, `rustc` itself is written in Rust. In principle, we could allow users to compile their Rust programs directly in the browser. We need a few additional features to achieve this objective, but it will happen in the near future.
```

Minor rewording for clarity

> Effectively _"supporting Rust"_ boils down to providing a Rust toolchain that can produce Wasm binaries for the BrowserPod target.

Reviewer: ```suggestion
Effectively _"supporting Rust"_ today boils down to providing a toolchain that can produce Wasm binaries for the BrowserPod target.
```

## Why not just pick WASI?

Our original plan was to implement a WASI (WebAssembly System Interface) layer on top of BrowserPod.
This solution would make any program targeting WASI immediately compatible with BrowserPod, whether it's written in Rust, Go, C, or anything else.

But when trying to compile existing Rust programs to WASI, we quickly realized that it's _not_ just a matter of switching build targets.

Take `yarn` for example. Compiling it for WASI would require:

- Platform-specific code to replace usage of `std::os::unix` with `std::os::wasi`: this is doable, although the mapping is not 1:1 (e.g. no absolute symlinks in WASI).
- Removal of thread usage: `wasm32-wasip3` has some limited cooperative threading support (without real parallelism), but it's very new and supported by neither the standard library nor [tokio](https://tokio.rs).
- Acceptance of reduced functionality: `yarn` relies on external programs for some functionality: in particular `git` (to fetch git dependencies) and `node` (to run lifecycle scripts) are required for basic operations.

> BrowserPod already has a solution to these problems, so we decided to skip the middleman

Reviewer: ```suggestion
BrowserPod already solves these problems, so we decided to skip the middleman
```

and implement our own Rust target: `wasm32-browserpod-linux-musl`.

## Making our own Rust target

Defining a [custom Rust target](https://doc.rust-lang.org/rustc/targets/custom.html) is _surprisingly easy_: you define your target's properties
in a .json file:

```json
{
  "llvm-target": "wasm32-unknown-unknown",
  "target-pointer-width": 32,
  "target-c-int-width": 32,
  "data-layout": "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20",
  "arch": "wasm64",
  "is-like-wasm": true,
  "target-family": ["unix"],
  "os": "linux",
  "env": "musl",
  "vendor": "browserpod",

  "linker-flavor": "wasm-ld",
  "linker": "rust-lld",
  ...
}
```

> And then pass the path of this file as the `--target` argument.

Reviewer: Reduce newlines here. Try to keep paragraphs about 3-4 sentences. Not too long, but one is too short

If you look closer at the json snippet of our target definition above, you will notice two interesting things:

- `target_family: ["unix"]`: originally, we had `["unix", "wasm"]`, but many crates use the wasm target family to mean "standalone web build", "WASI", "no filesystem", or similar. Dropping "wasm" seems to be _completely harmless_.

- `arch: "wasm64"`: this is even more surprising, but unfortunately also necessary. `arch="wasm32"` is often used to detect standalone web targets (for example, [`reqwest`](https://docs.rs/reqwest) will replace the whole HTTP client with a `fetch()` implementation). You would think that something would break because of this, but important info like the size of pointers is defined separately, and the actual target passed to the LLVM backend is still `wasm32`, so it all works out!

Of course, we would love to get rid of these hacks, but the proper solution requires awareness in the ecosystem that fully-featured Wasm targets exist, which will take time (but we have a [plan](#future-plans)).

## The payoff

The original goal was to run `yarn`, but since our approach was completely general, we found that
a lot of programs _just work_.

> Here is a (non-exhaustive) list of popular programs that I tried (but we expect most CLI applications to just work):

Reviewer: ```suggestion
Here is a (non-exhaustive) list of popular programs that I tried, but we expect most CLI applications to just work:
```

| [yarnpkg/zpm](https://github.com/yarnpkg/zpm)               | ~50k       | **no changes**                   | _working_ |
| [starship/starship](https://github.com/starship/starship)   | ~50k       | **no changes**                   | _working_ |
| [jj-vcs/jj](https://github.com/jj-vcs/jj)                   | ~200k      | **no changes**                   | _working_ |
| [openai/codex](https://github.com/openai/codex)             | ~1.25M     | disabled some sandboxing options | _working_ |

You can play with them in the terminal below:

<BrowserPodTerminal
	client:only="svelte"
	ctxId="bp-rust-demo"
	description="A playground with a collection of rust programs"
	autoStart={true}
	height="18rem"
	terminalTabs={[
		{
			id: "demo2",
			label: "demo2",
			echo: false,
			command: "bash",
			args: [
				"-c",
				"set -x; yarn -v; rg --version; jj --version; starship --version; exec bash",
			],
			env: ["PATH=/home/.bin:/bin"],
		},
	]}
/>

Codex is fairly large and requires an API key to run, so you can also see it in action in
the following video:

<video controls autoplay loop muted playsinline>

> 	<source src="/blog/bp-rust/codex.mp4" type="video/mp4" />

Reviewer: Video does not load, unsure if it's a preview issue

Author: not made yet

</video>

## Using it

Rust has a single blessed package manager and build system: `cargo`. This makes it much easier for us to provide a good out-of-the-box developer experience compared to, say, C/C++.

Rust also has an official way to obtain and manage multiple toolchains: **[rustup](https://rustup.rs)**. We can't be distributed _through_ rustup yet, because we are not an officially supported target, but we can install into its directories, so you can enable our toolchain for a single project in the usual way:

```bash
# manually install our toolchain (one day this could be via rustup too)
curl https://rt.browserpod.io/3.0.0/rust/install.sh | bash

# override the toolchain for the current project with rustup
rustup override set browserpod-3.0.0

# build
cargo build --target wasm32-browserpod-linux-musl

# uninstall via rustup
rustup toolchain uninstall browserpod-3.0.0
```

## Try it now

Head to [console.browserpod.io](https://console.browserpod.io) to get an API key. You can find the documentation on [browserpod.io/docs](https://browserpod.io/docs).

> If you have a Rust project you'd like to see running in a browser tab — especially one you expect to break — we'd genuinely like to hear about it. Come find us on [Discord](https://discord.leaningtech.com).

Reviewer: I see evil em-dashes here?

Author: :man_facepalming: they sneaked in in a grammar pass with Claude (I already converted them to normal "-", but better just to remove)

> ## Future plans

Reviewer: Maybe swap "Try it now" and "Future plans"?

	version="3.0.0"
	projectSource={{ type: "local", path: "bp-rust" }}
	userImage="wss://disks.browserpod.io/rust-post-demos-2.ext2"

> 	apiKey="bp1_ff7fc043f1736f5a7463a786c8983c68c53a00e4bfa3b704275e6937446d3522"

Reviewer: And we might have some dev keys to purge now

Author: yep, it was not meant to be committed...
